### ISSUE TITLE
[network] Improve initialization via NetworkBuilder

### ISSUE THREAD

**Issue by bothra90** [libra/libra#3342]
The `NetworkBuilder` module has seen significant change over time which give it a "batteries included, but optional" nature, since modules such as Discovery and HealthChecker are only added if needed. However, there's still room for some simplifications:
  - arguments (such as `trusted_peers`) that are specific to `discovery` module should be initialized in the `add_discovery` function rather than separately, which leaves room for bugs.
  - Similarly, arguments for health_checker (`ping_interval_ms` and `ping_timeout_ms`) should be sent in `add_connection_monitoring`.

**Comment by davidiw** [comment]
Before jumping into there, we should consider merging NetworkBuilder and NetworkConfig, since the two are basically the same thing.

**Comment by Rustinante** [comment]
Agreed, the `NetworkBuilder` seems to have too many states, refactoring the code into a more functional style probably be better.

**Comment by Rustinante** [comment]
Hi @davidiw and @phlip9, I've created a PR #3881 for this issue to refactor the `NetworkBuilder`. It turns out we need a structure to store the various network senders, and it feels a little weird to put them into the `NetworkConfig`, so I created another struct called `NodeNetwork` to store them, so now we have these two structures, where the first one is the configs responsible to set up the network, while the second one is sort of the result of setting up the network...

Please let me know what you think, and we can refactor this further.
